### PR TITLE
Reduce NetworKit's dependencies

### PR DESCRIPTION
**Pull request by cndolo**
In this PR we try to reduce NetworKit's Python dependencies by making most packages optional.
If the package is not available when it is needed, a runtime error is thrown. 
The runtime error, `MissingDependencyError` specifies which particular package was needed.

### CODE REVIEW

**Comment by avdgrinten** [comment]
One issue here is that notebooks (e.g., UserGuide) use optional dependencies (such as matplotlib). We need to work around that somehow.

**Comment by cndolo** [comment]
> One issue here is that notebooks (e.g., UserGuide) use optional dependencies (such as matplotlib). We need to work around that somehow.

Additionally, almost all (if not all) the Python tests use optional dependencies.

File: .travis.yml [hunk 1]
```diff
@@ -305,6 +305,8 @@ script:
  - export LD_LIBRARY_PATH=${LD_LIBRARY_PATH:+$LD_LIBRARY_PATH:}$(pwd)/core_build
  - NETWORKIT_PARALLEL_JOBS=2 python3 ./setup.py build_ext --inplace --networkit-external-core
  - NETWORKIT_PARALLEL_JOBS=2 pip3 install -e .
+ - python3 -c "import networkit"
```
Comment (avdgrinten) [inline .travis.yml]: Please add a comment on why we add this line; it is not immediately obvious.

File: benchmark/Benchmark.py [hunk 2]
```diff
@@ -8,40 +8,51 @@
...
 from util import *
+from networkit.support import MissingDependencyError
 import nk
 
 try:
 	import networkx
 	import nx
 except ImportError as ex:
-	print("networkx not available")
+	networkx = None
```
Comment (avdgrinten) [inline benchmark/Benchmark.py]: What about the `nx` variable here? We should either consistently define all module symbols as `None` if the packages are not available, or consistently only set one symbol per package to `None`.

File: benchmark/Benchmark.py [hunk 3]
```diff
@@ -8,40 +8,51 @@
...
 
 try:
 	import igraph
 	import ig
 except ImportError as ex:
-	print("igraph not available")
+	igraph = None
+	ig = None
```
Comment (avdgrinten) [inline benchmark/Benchmark.py]: For example, here we set both `igraph` and `ig` to `None`.

File: benchmark/Benchmark.py [hunk 4]
```diff
@@ -8,40 +8,51 @@
...
 import sys
 import warnings
 import math
 import os
 import numpy
-import matplotlib.pyplot as plt
-import seaborn
+try:
+	import matplotlib.pyplot as plt
+except ImportError as ex:
```
Comment (avdgrinten) [inline benchmark/Benchmark.py]: The exception `ex` is not needed, right? Do not give it a name. (This applies to the other `except` statements, too.)

File: benchmark/Benchmark.py (hunk 2 above)
Comment (avdgrinten) [reply benchmark/Benchmark.py]: An alternative would be to have a Boolean `have_networkx` variable that we set to `False` if the import failed. Maybe that is the cleanest solution.

File: networkit/graphio.py [hunk 5]
```diff
@@ -10,19 +10,17 @@
...
-except ImportError:
-	print("module 'numpy' not available - some functionality will be restricted")
-try:
-	import scipy.io
-except ImportError:
-	print("module 'scipy' not available - some functionality will be restricted")
-import fnmatch
+	import fnmatch
+except ImportError as importError:
+	fnmatch = None
```
Comment (avdgrinten) [inline networkit/graphio.py]: `fnmatch` is part of the Python standard library, so it must always be available.

File: networkit/profiling/stat.py [hunk 6]
```diff
@@ -514,9 +519,9 @@ def funcHexBinning(sample_1, sample_2):
...
 			plt.close(fig)
 			return result
-		results["Binning"] = funcHexBinning(
-			sample_1,
-			sample_2
-		) 
+			results["Binning"] = funcHexBinning(
+				sample_1,
+				sample_2
+			)
```
Comment (avdgrinten) [inline networkit/profiling/stat.py]: This looks like it changes the logic. Is that intended?

**Comment by cndolo** [comment]
Kindly review the new changes as per what we discussed.

**Pull request merged**